samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -0,0 +1,166 @@
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.zkLeaderElectionListener = leaderElectionListener;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  public ZkLeaderElector(String processorIdStr, ZkUtils zkUtils) {
+    this.zkLeaderElectionListener = new ZkLeaderElectionListener();
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  // TODO: This should go away once we integrate with Zk based Job Coordinator
+  private String getHostName() {
+    try {
+      return InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      LOGGER.error("Failed to fetch hostname of the processor", e);
+      throw new SamzaException(e);
+    }
+  }
+
+  @Override
+  public boolean tryBecomeLeader() {
+    String currentPath = zkUtils.getEphemeralPath();
+
+    if (currentPath == null || currentPath.isEmpty()) {
+      currentPath = zkUtils.registerProcessorAndGetId(hostName);
+    }
+
+    List<String> children = zkUtils.getActiveProcessors();
+    LOGGER.debug(zLog("Current active processors - " + children));
+    int index = children.indexOf(ZkKeyBuilder.parseIdFromPath(currentPath));
+
+    if (children.size() == 0 || index == -1) {
+      throw new SamzaException("Looks like we are no longer connected to Zk. Need to reconnect!");
+    }
+
+    leaderId = ZkKeyBuilder.parseIdFromPath(children.get(0));
+    if (index == 0) {
+      LOGGER.info(zLog("Eligible to become the leader!"));
+      return true;
+    }
+
+    LOGGER.info("Index = " + index + " Not eligible to be a leader yet!");
+    String predecessor = children.get(index - 1);
+    if (!predecessor.equals(currentSubscription)) {
+      if (currentSubscription != null) {
+        LOGGER.debug(zLog("Unsubscribing data change for " + currentSubscription));
+        zkUtils.unsubscribeDataChanges(keyBuilder.getProcessorsPath() + "/" + currentSubscription, zkLeaderElectionListener);
+      }
+      currentSubscription = predecessor;
+      LOGGER.info(zLog("Subscribing data change for " + predecessor));
+      zkUtils.subscribeDataChanges(keyBuilder.getProcessorsPath() + "/" + currentSubscription, zkLeaderElectionListener);
+    }
+    /**
+     * Verify that the predecessor still exists. This step is needed because the ZkClient subscribes for data changes
+     * on the path, even if the path doesn't exist. Since we are using Ephemeral Sequential nodes, if the path doesn't
+     * exist during subscription, it is not going to get created in the future.
+     */
+    boolean predecessorExists = zkUtils.exists(keyBuilder.getProcessorsPath() + "/" + currentSubscription);
+    if (predecessorExists) {
+      LOGGER.info(zLog("Predecessor still exists. Current subscription is valid. Continuing as non-leader."));
+    } else {
+      try {
+        Thread.sleep(random.nextInt(1000));
```

Comment:
No reason. In a flaky system, if there are many processors trying the leader election at the same time, it will cause a higher load on ZK than normal. This kind of mitigates that issue slightly by doing some random back-off before retry.